File: webapp/frontend/src/types/event.ts
// ============================================
// Event Types
// ============================================

/**
 * Event type enum - matches backend oneof validation
 */
export type EventType =
  | "historical_event"
  | "national_day"
  | "world_day"
  | "anniversary"
  | "cultural"
  | "military";

/**
 * Event importance enum
 */
export type EventImportance = "high" | "medium" | "low";

/**
 * Event data returned from the API (matches EventResponse DTO)
 */
export interface HistoricalEvent {
  id: string;
  title: string;
  slug: string;
  event_date: string;
  event_day: number;
  event_month: number;
  event_year?: number | null;
  is_lunar: boolean;
  is_recurring: boolean;
  event_type: EventType;
  country: string;
  country_code?: string;
  flag_emoji?: string;
  short_description: string;
  image_url: string;
  article_id: string | null;
  importance: EventImportance;
  tags: string[];
  is_active: boolean;
  created_at: string;
}

/**
 * Event summary (for list views)
 */
export interface EventSummary {
  id: string;
  title: string;
  slug: string;
  event_date: string;
  event_day: number;
  event_month: number;
  event_type: EventType;
  importance: EventImportance;
  country: string;
  image_url?: string;
  is_lunar: boolean;
  is_active: boolean;
  created_at: string;
}

/**
 * Create event request payload (matches CreateEventRequest DTO)
 */
export interface CreateEventRequest {
  title: string;
  event_date?: string;
  event_day: number;
  event_month: number;
  event_year?: number;
  is_lunar?: boolean;
  is_recurring?: boolean;
  event_type: string;
  country?: string;
  country_code?: string;
  flag_emoji?: string;
  short_description?: string;
  image_url?: string;
  article_id?: string;
  importance?: string;
  tags?: string[];
}

/**
 * Update event request payload (matches UpdateEventRequest DTO)
 */
export interface UpdateEventRequest {
  title?: string;
  event_date?: string;
  event_day?: number;
  event_month?: number;
  event_year?: number;
  is_lunar?: boolean;
  is_recurring?: boolean;
  event_type?: string;
  country?: string;
  country_code?: string;
  flag_emoji?: string;
  short_description?: string;
  image_url?: string;
  article_id?: string;
  importance?: string;
  tags?: string[];
}

/**
 * Event list query parameters
 */
export interface EventListParams {
  page?: number;
  limit?: number;
  search?: string;
  sort_by?: string;
  sort_order?: "asc" | "desc";
  event_type?: string;
  importance?: string;
  country?: string;
}

File: webapp/backend/internal/models/event.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/lib/pq"
	"gorm.io/gorm"
)

// EventType constants
const (
	EventTypeHistorical  = "historical_event"
	EventTypeNationalDay = "national_day"
	EventTypeWorldDay    = "world_day"
	EventTypeAnniversary = "anniversary"
	EventTypeCultural    = "cultural"
	EventTypeMilitary    = "military"
)

// EventImportance constants
const (
	EventImportanceLow    = "low"
	EventImportanceMedium = "medium"
	EventImportanceHigh   = "high"
)

// Event represents a historical event, national day, or international day.
type Event struct {
	ID               uuid.UUID      `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	Title            string         `gorm:"type:varchar(500);not null" json:"title"`
	Slug             string         `gorm:"type:varchar(500);not null;uniqueIndex" json:"slug"`
	EventDate        *time.Time     `gorm:"type:date" json:"event_date,omitempty"`
	EventDay         int            `gorm:"type:int;not null" json:"event_day"`
	EventMonth       int            `gorm:"type:int;not null" json:"event_month"`
	EventYear        *int           `gorm:"type:int" json:"event_year,omitempty"`
	IsLunar          bool           `gorm:"default:false" json:"is_lunar"`
	IsRecurring      bool           `gorm:"default:true" json:"is_recurring"`
	EventType        string         `gorm:"type:varchar(50);not null;default:'historical_event'" json:"event_type"`
	Country          string         `gorm:"type:varchar(100)" json:"country,omitempty"`
	CountryCode      string         `gorm:"type:varchar(5)" json:"country_code,omitempty"`
	FlagEmoji        string         `gorm:"type:varchar(10)" json:"flag_emoji,omitempty"`
	ShortDescription string         `gorm:"type:text" json:"short_description,omitempty"`
	ImageURL         string         `gorm:"type:varchar(500)" json:"image_url,omitempty"`
	ArticleID        *uuid.UUID     `gorm:"type:uuid;index" json:"article_id,omitempty"`
	Importance       string         `gorm:"type:varchar(20);not null;default:'medium'" json:"importance"`
	Tags             pq.StringArray `gorm:"type:text[];default:'{}'" json:"tags,omitempty"`
	IsActive         bool           `gorm:"default:true" json:"is_active"`
	CreatedAt        time.Time      `gorm:"type:timestamptz;not null;default:now()" json:"created_at"`
	UpdatedAt        time.Time      `gorm:"type:timestamptz;not null;default:now()" json:"updated_at"`
	DeletedAt        gorm.DeletedAt `gorm:"type:timestamptz;index" json:"deleted_at,omitempty"`

	// Associations
	Article *Article `gorm:"foreignKey:ArticleID" json:"article,omitempty"`
}

func (Event) TableName() string {
	return "events"
}
